Use 32bit relative pointers in ExternalReferenceTable

MichalStrehovsky opened the pull request:
We can't use IMAGE_REL_BASED_ADDR32NB (because it's Windows only), but
we can use IMAGE_REL_BASED_REL32 which is close enough.

For CppCodegen we'll need to stick to pointer-sized entries (that are
bigger and require an extra reloc), but CppCodegen is not far enough to
use this yet anyway. We can cross that bridge when we get there.

We might also just want to just standardize on IMAGE_REL_BASED_REL32 for
any TypeManager-based modules for simplicity.

This drops the size of HelloWorld.exe with shared generics enabled from 4,390,912 bytes to 4,297,216 bytes.

jkotas commented on src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExternalReferencesTableNode.cs:
```diff
@@ -97,14 +97,14 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
                 {
                     // TODO: set low bit if the linkage of the symbol is IAT_PVALUE.
-                    builder.EmitPointerReloc(symbolAndDelta.Symbol, symbolAndDelta.Delta);
+                    builder.EmitReloc(symbolAndDelta.Symbol, RelocType.IMAGE_REL_BASED_REL32, symbolAndDelta.Delta);
```
I think it would look better to use `IMAGE_REL_BASED_RELPTR32` here instead.

To save the hassle with updating the ObjectWriter, it can be special cased in EmitSymbolRef for now: map it to `IMAGE_REL_BASED_REL32` and add sizeof(int) to delta.

jkotas commented on src/System.Private.TypeLoader/src/Internal/Runtime/TypeLoader/ExternalReferencesTable.cs:
```diff
@@ -95,8 +91,9 @@ unsafe public IntPtr GetIntPtrFromIndex(uint index)
+            int* elementsTable = (int*)_elements;
 
-            return ((TableElement*)_elements)[index];
+            return (IntPtr)((byte*)&elementsTable[index + 1] + elementsTable[index]);
```
With `IMAGE_REL_BASED_REL32`, this can be something like:

```
int* pRelPtr32 = ((int*)_elements[index]);
return (IntPtr)((byte*)pRelPtr32 + *pRelPtr32);
```

MichalStrehovsky commented:
> I think it would look better to use IMAGE_REL_BASED_RELPTR32 here instead.

Yup, looks better.

I fixed it up and fixed a few other places that should use that instead.

Cc @nattress @davidwrighton

MichalStrehovsky commented:
@davidwrighton I validated this shouldn't cause problems on the TFS side. We already use `IMAGE_REL_BASED_RELPTR32` on `target.Abi == TargetAbi.ProjectN` so this won't be new.
